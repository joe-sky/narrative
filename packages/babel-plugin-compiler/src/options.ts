import type { BabelFile } from '@babel/core';

export type State = {
  get: (name: string) => any;
  set: (name: string, value: any) => any;
  opts: CompilerOptions;
  file: BabelFile;
};

export interface CompilerOptions {
  importedLib?: string[] | 'none';
}
