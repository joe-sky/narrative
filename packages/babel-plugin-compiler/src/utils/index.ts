import * as BabelCore from '@babel/core';

export const NT_CONTROL_FLOW = '@narrative/control-flow';

export type State = {
  get: (name: string) => any;
  set: (name: string, value: any) => any;
  opts: CompilerOptions;
  file: BabelCore.BabelFile;
};

export interface CompilerOptions {
  importedLib?: string[] | 'none';
}
