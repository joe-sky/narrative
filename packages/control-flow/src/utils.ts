export type JSXNode = JSX.Element | string | number | boolean | null | undefined | Record<any, any>;

export type JSXChild = JSXNode | JSXNode[];

export interface Childrenable {
  children?: JSXChild;
}
