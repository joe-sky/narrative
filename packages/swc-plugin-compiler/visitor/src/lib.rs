mod tags;
mod utils;
pub mod options;
pub mod visitor;