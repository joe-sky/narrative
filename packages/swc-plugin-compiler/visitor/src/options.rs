use serde::Deserialize;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct Options {
  pub imported_lib: Option<Vec<String>>,
}

impl Default for Options {
  fn default() -> Self {
    Self {
      imported_lib: None,
    }
  }
}