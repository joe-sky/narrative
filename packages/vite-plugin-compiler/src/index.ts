import { Plugin } from 'vite';
import { transformSync, ParserOptions } from '@babel/core';
import babelPluginNt from '@narrative/babel-plugin-compiler';

export interface Options {
  parserPlugins?: ParserOptions['plugins'];
}

export default function ntPlugin(opts?: Options) {
  return {
    name: 'narrative',

    enforce: 'pre',

    transform(code, id) {
      if (!/\.(t|j)sx?$/.test(id) || id.includes('node_modules')) {
        return;
      }

      // plain js/ts files can't use jsx-sfc without importing it, so skip
      // them whenever possible
      if (!id.endsWith('x') && !code.includes('@narrative/control-flow')) {
        return;
      }

      const parserPlugins: ParserOptions['plugins'] = [
        'jsx',
        'importMeta',
        // since the plugin now applies before esbuild transforms the code,
        // we need to enable some stage 3 syntax since they are supported in
        // TS and some environments already.
        'topLevelAwait',
        'classProperties',
        'classPrivateProperties',
        'classPrivateMethods'
      ];
      if (/\.tsx?$/.test(id)) {
        // it's a typescript file
        // TODO: maybe we need to read tsconfig to determine parser plugins to
        // enable here, but allowing decorators by default since it's very
        // commonly used with TS.
        parserPlugins.push('typescript', 'decorators-legacy');
      }
      if (opts?.parserPlugins) {
        parserPlugins.push(...opts.parserPlugins);
      }

      const result = transformSync(code, {
        configFile: false,
        filename: id,
        parserOpts: {
          sourceType: 'module',
          allowAwaitOutsideFunction: true,
          plugins: parserPlugins
        },
        plugins: [babelPluginNt],
        sourceMaps: true,
        sourceFileName: id
      });

      if (result?.code == null) {
        // no component detected in the file
        return code;
      }

      return {
        code: result.code,
        map: result.map
      };
    }
  } as Plugin;
}
