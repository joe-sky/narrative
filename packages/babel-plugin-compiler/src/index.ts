import jsx from '@babel/plugin-syntax-jsx';
import { Visitor, NodePath } from '@babel/traverse';
import * as types from '@babel/types';
import * as astUtil from './utils/ast';
import { transformIf, SUB_TAGS_IF } from './tags/if';
import { transformSwitch, SUB_TAGS_SWITCH } from './tags/switch';
import { transformFor, SUB_TAGS_FOR } from './tags/for';
import { State, CompilerOptions } from './utils';

export { CompilerOptions };

const nodeTransforms = {
  If: transformIf,
  Switch: transformSwitch,
  For: transformFor
};

export default function NtCompiler() {
  return {
    inherits: jsx,
    name: '@narrative/babel-plugin-compiler',
    visitor: {
      JSXElement: {
        enter(path, state: State) {
          const elName = path.node.openingElement.name;
          if (!types.isJSXIdentifier(elName)) {
            return;
          }

          const nodeName = elName.name;
          const { importedLib } = state?.opts;
          if (importedLib !== 'none' && !astUtil.isImportedByLib(nodeName, path, importedLib)) {
            return;
          }

          const transform = nodeTransforms[nodeName as keyof typeof nodeTransforms];
          if (transform) {
            path.replaceWith(transform(path.node));
          }
        }
      },
      Program(_path, state: State) {
        _path.traverse({
          JSXElement: {
            enter(path) {
              if (!types.isJSXElement(path.parent)) {
                return;
              }

              const elName = path.node.openingElement.name;
              if (!types.isJSXIdentifier(elName)) {
                return;
              }

              const nodeName = elName.name;
              if (
                !astUtil.isImportedByLib(nodeName, path, state?.opts.importedLib) ||
                SUB_TAGS_IF[nodeName as keyof typeof SUB_TAGS_IF] ||
                SUB_TAGS_SWITCH[nodeName as keyof typeof SUB_TAGS_SWITCH] ||
                SUB_TAGS_FOR[nodeName as keyof typeof SUB_TAGS_FOR]
              ) {
                return;
              }

              path.replaceWith(types.jsxExpressionContainer(path.node));
            }
          }
        });
      }
    } as Visitor
  };
}
