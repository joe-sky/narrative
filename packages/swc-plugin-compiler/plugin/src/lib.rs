#![allow(clippy::not_unsafe_ptr_arg_deref)]

use narrative_swc_compiler_visitor::visitor;
use swc_core::{
  ecma::{ ast::Program, visit::FoldWith },
  plugin::{ plugin_transform, proxies::TransformPluginProgramMetadata },
};
use visitor::transform_narrative;

#[plugin_transform]
pub fn process_transform(program: Program, metadata: TransformPluginProgramMetadata) -> Program {
  let options = metadata
    .get_transform_plugin_config()
    .map(|json| { serde_json::from_str(&json).expect("failed to get plugin config for Narrative") })
    .unwrap_or_default();

  program.fold_with(&mut transform_narrative(options, metadata.unresolved_mark, metadata.comments))
}