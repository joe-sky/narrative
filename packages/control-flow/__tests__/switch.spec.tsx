import React, { useState, useEffect } from 'react';
import { render } from '@testing-library/react';
import { Switch, Case, Default } from '../src/index';

interface Props {
  expression?: number;
  expressionStr?: string;
}

const TestSwitch: React.FC<Props> = props => {
  return (
    <Switch value={props.expression}>
      <Case is={1}>
        <i>test1</i>
      </Case>
      <Case is={2}>
        <i>test2</i>
      </Case>
      <Default>
        <i>test3</i>
      </Default>
    </Switch>
  );
};

test('switch-case with value', function () {
  const el = render(<TestSwitch expression={2} />);
  expect(el.container.firstChild).toMatchSnapshot(`<i>test2</i>`);
});

const TestSwitchValues: React.FC<Props> = props => {
  return (
    <Switch value={props.expressionStr}>
      <Case is="1">
        <i>test1</i>
      </Case>
      <Case in={['2', '3']}>
        <i>test2</i>
      </Case>
      <Default>
        <i>test3</i>
      </Default>
    </Switch>
  );
};

test('switch-case with values', function () {
  const el = render(<TestSwitchValues expressionStr="2" />);
  expect(el.container.firstChild).toMatchSnapshot(`<i>test2</i>`);
});
