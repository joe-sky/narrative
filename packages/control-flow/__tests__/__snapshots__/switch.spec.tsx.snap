// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switch-case with value: <i>test2</i> 1`] = `
<i>
  test2
</i>
`;

exports[`switch-case with values: <i>test2</i> 1`] = `
<i>
  test2
</i>
`;
